// jant-taf/src/component/edit/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.edit = void 0;
const schematics_1 = require("@angular-devkit/schematics");
const strings_1 = require("@angular-devkit/core/src/utils/strings");
// You don't have to export the function as default. You can also have more than one rule factory
// per file.
function edit(options) {
    const { table, projectName, parent, tafConfig } = options;
    const componentType = "edit";
    const componentName_class = (0, strings_1.classify)(componentType) + (0, strings_1.classify)(table.table); //"ListLivre"
    const componentName = (0, strings_1.dasherize)(componentName_class);
    const fullComponentName = parent + (0, strings_1.dasherize)(table.table) + "/" + componentName; // /home/auteur/list-auteur
    return (tree) => {
        // Le composant existe dèja
        let c_path = `/src/app/${fullComponentName}/${componentName}.component.ts`;
        if (tree.exists(c_path)) {
            console.log(`Le composant ${componentName} existe dèja dans le module ${parent} : ${c_path}`);
            return tree;
        }
        // le composant n'existe pas encore
        return (0, schematics_1.chain)([
            // création du composant
            (0, schematics_1.externalSchematic)("@schematics/angular", "component", { project: projectName, name: fullComponentName, style: tafConfig.style }),
            // modification des fichiers
            (tree, _context) => {
                // modifier le fichier HTML
                tree.overwrite(`src/app/${fullComponentName}/${componentName}.component.html`, // home/auteur/lis-auteur/list-auteur.component.html
                get_html_content(table));
                // modifier le fichier TS
                tree.overwrite(`src/app/${fullComponentName}/${componentName}.component.ts`, // home/auteur/lis-auteur/list-auteur.component.ts
                get_ts_content(table, componentName_class, componentName, tafConfig));
            }
        ]);
    };
}
exports.edit = edit;
function get_ts_content(table, componentName_class, componentName, tafConfig) {
    let validators_update = table.table_descriptions.les_colonnes.map((une_colonne) => {
        if (une_colonne["Key"] != 'PRI' && !(une_colonne["Field"] == "created_at" && une_colonne["Default"] != "")) {
            if (une_colonne["Null"] == "NO") {
                return `${une_colonne["Field"]} : [${table.table}_to_edit.${une_colonne["Field"]}, Validators.required]`;
            }
            else {
                return `${une_colonne["Field"]} : [${table.table}_to_edit.${une_colonne["Field"]}]`;
            }
        }
        return undefined;
    }).filter((une_colonne) => une_colonne != undefined).join(",\n");
    table.description.map((une_colonne) => {
        return `${une_colonne}: ["", Validators.required]`;
    }).join(",\n");
    return `
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { FormGroup, FormBuilder, Validators } from '@angular/forms';
import { ApiService } from '../../../service/api/api.service';
@Component({
  selector: 'app-${componentName}',
  templateUrl: './${componentName}.component.html',
  styleUrls: ['./${componentName}.component.${tafConfig.style}']
})
export class ${componentName_class}Component {
  reactiveForm_edit_${table.table} !: FormGroup;
  submitted: boolean = false
  loading_edit_${table.table}: boolean = false
  @Input()
  ${table.table}_to_edit: any = {}
  @Output()
  cb_edit_${table.table}=new EventEmitter()
  form_details: any = {}
  loading_get_details_add_${table.table}_form = false
  constructor(private formBuilder: FormBuilder, public api: ApiService) { 
      
  }
  ngOnInit(): void {
      this.get_details_add_${table.table}_form()
      this.update_form(this.${table.table}_to_edit)
  }
  // mise à jour du formulaire
  update_form(${table.table}_to_edit:any) {
      this.reactiveForm_edit_${table.table} = this.formBuilder.group({
          ${validators_update}
      });
  }

  // acces facile au champs de votre formulaire
  get f(): any { return this.reactiveForm_edit_${table.table} .controls; }
  // validation du formulaire
  onSubmit_edit_${table.table}() {
      this.submitted = true;
      console.log(this.reactiveForm_edit_${table.table}.value)
      // stop here if form is invalid
      if (this.reactiveForm_edit_${table.table}.invalid) {
          return;
      }
      var ${table.table} = this.reactiveForm_edit_${table.table}.value
      this.edit_${table.table}({
      condition:JSON.stringify({id_${table.table}:this.${table.table}_to_edit.id_${table.table}}),
      data:JSON.stringify(${table.table})
      })
  }
  // vider le formulaire
  onReset_edit_${table.table}() {
      this.submitted = false;
      this.reactiveForm_edit_${table.table}.reset();
  }
  edit_${table.table}(${table.table}: any) {
      this.loading_edit_${table.table} = true;
      this.api.taf_post("${table.table}/edit", ${table.table}, (reponse: any) => {
          if (reponse.status) {
              this.cb_edit_${table.table}.emit({
                  new_data:JSON.parse(${table.table}.data)
              })
              console.log("Opération effectuée avec succés sur la table ${table.table}. Réponse= ", reponse);
              this.onReset_edit_${table.table}()
              alert("Opération effectuée avec succés sur la table ${table.table}")
          } else {
              console.log("L\'opération sur la table ${table.table} a échoué. Réponse= ", reponse);
              alert("L'opération a echoué")
          }
          this.loading_edit_${table.table} = false;
      }, (error: any) => {
          this.loading_edit_${table.table} = false;
      })
  }
  get_details_add_${table.table}_form() {
      this.loading_get_details_add_${table.table}_form = true;
      this.api.taf_post("${table.table}/get_form_details", {}, (reponse: any) => {
        if (reponse.status) {
          this.form_details = reponse.data
          console.log("Opération effectuée avec succés sur la table ${table.table}. Réponse= ", reponse);
        } else {
          console.log("L'opération sur la table ${table.table} a échoué. Réponse= ", reponse);
          alert("L'opération a echoué")
        }
        this.loading_get_details_add_${table.table}_form = false;
      }, (error: any) => {
      this.loading_get_details_add_${table.table}_form = false;
    })
  }
}
`;
}
function get_html_content(table) {
    let all_colonne = table.table_descriptions.les_colonnes.map((une_colonne) => {
        switch (une_colonne.Key) {
            case "PRI":
                break;
            case "MUL":
                return `<!-- champs ${une_colonne.Field} avec un control de validite : clé étrangère liée à la colonne ${une_colonne.Field} de la table ${une_colonne.referenced_table.table_name} -->
        <div class="form-group col-sm-6">
          <label >${une_colonne.referenced_table.table_name}</label>
          <select [ngClass]="{ 'is-invalid': submitted && f.${une_colonne.Field}.errors }" class="form-select" formControlName="${une_colonne.Field}">
            <option value="">Sélectionnez un(e) ${une_colonne.referenced_table.table_name}</option>
            <option [value]="one_${une_colonne.referenced_table.table_name}.${une_colonne.referenced_table.cle_primaire.Field}" *ngFor="let one_${une_colonne.referenced_table.table_name} of form_details.les_${une_colonne.referenced_table.table_name}s">{{"${une_colonne.referenced_table.table_name} N°"+one_${une_colonne.referenced_table.table_name}.${une_colonne.referenced_table.cle_primaire.Field}}}</option>
          </select>
          <div *ngIf="submitted && f.${une_colonne.Field}.errors" class="invalid-feedback">
            <div *ngIf="f.${une_colonne.Field}.errors.required">ce champ est obligatoire</div>
          </div>
        </div>`;
            default:
                if (une_colonne["Field"] == "created_at" && une_colonne["Default"] == "CURRENT_TIMESTAMP") {
                    return;
                }
                else {
                    return `<!-- ${une_colonne.Field} field avec un control de validite -->
          <div class="form-group col-sm-6">
          <label >${une_colonne.Field}</label>
          <input class="form-control" type="text"  formControlName="${une_colonne.Field}"  placeholder="${une_colonne.Field}"  [ngClass]="{ 'is-invalid': submitted && f.${une_colonne.Field}.errors }"/>
          <div *ngIf="submitted && f.${une_colonne.Field}.errors" class="invalid-feedback">
          <div *ngIf="f.${une_colonne.Field}.errors.required"> ${une_colonne.Field} est obligatoire </div>
          </div>
          </div>  `;
                }
        }
    })
        .join("\n");
    return `
<form  [formGroup]="reactiveForm_edit_${table.table} " (ngSubmit)="onSubmit_edit_${table.table} ()" #form_edit_${table.table} ="ngForm" class="row">
  ${all_colonne}
</form>
<!-- vous pouvez valider votre formulaire n\'importe ou -->

<div class="text-center m-2">
    <button type="button" class="btn btn-primary m-2" [disabled]="loading_edit_${table.table} "
        (click)="form_edit_${table.table} .ngSubmit.emit()">{{loading_edit_${table.table} ?"En cours ...":"Valider"}}</button>
    <button class="btn btn-secondary m-2" type="reset" (click)="onReset_edit_${table.table} ()">Vider</button>
</div>
`;
}
